sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+use crate::io::spi::Target;
+use crate::io::uart::Uart;
+use crate::transport::{Capabilities, Capability, Transport, TransportError};
+use crate::util::usb::UsbBackend;
+
+pub mod gpio;
+pub mod uart;
+pub mod spi;
+
+/// Implementation of the Transport trait for HyperDebug based on the
+/// Nucleo-L552ZE-Q.
+pub struct Hyperdebug {
+    spi_names: HashMap<String, u8>,
+    spi_interface: BulkInterface,
+    uart_ttys: HashMap<String, PathBuf>,
+    inner: Rc<Inner>,
+}
+
+/// Index of a single USB "interface", with its associated IN and OUT
+/// endpoints.  Used to instantiate e.g. SPI trait.
+#[derive(Copy, Clone)]
+pub struct BulkInterface {
+    interface: u8,
+    in_endpoint: u8,
+    out_endpoint: u8,
+}
+
+/// Internal state of the Hyperdebug struct, this struct is reference counted such that Gpio,
+/// Spi and Uart sub-structs can all refer to this shared data, which is guaranteed to live on,
+/// even if the caller lets the outer Hyperdebug struct run out of scope.
+pub struct Inner {
+    console_tty: PathBuf,
+    usb_device: RefCell<UsbBackend>,
+    gpio: RefCell<HashMap<String, Rc<dyn GpioPin>>>,
+    spis: RefCell<HashMap<u8, Rc<dyn Target>>>,
+    uarts: RefCell<HashMap<PathBuf, Rc<dyn Uart>>>,
+}
+
+impl Hyperdebug {
+    pub const VID_GOOGLE: u16 = 0x18d1;
+    pub const PID_HYPERDEBUG: u16 = 0x520e;
+
+    /// Establish connection with a particular HyperDebug.
+    pub fn open(usb_vid: Option<u16>, usb_pid: Option<u16>, usb_serial: &Option<String>)
+                -> Result<Self> {
+        let device = UsbBackend::new(
+            usb_vid.unwrap_or(Self::VID_GOOGLE),
+            usb_pid.unwrap_or(Self::PID_HYPERDEBUG),
+            usb_serial)?;
+
+        let path = PathBuf::from("/sys/bus/usb/devices");
+
+        let mut console_tty: Option<PathBuf> = None;
+        let mut spi_interface: Option<BulkInterface> = None;
+        let mut uart_ttys: HashMap<String, PathBuf> = HashMap::new();
+
+        let config_desc = device.active_config_descriptor()?;
+        // Iterate through each USB interface, discovering e.g. supported UARTs.
+        for interface in config_desc.interfaces() {
+            for interface_desc in interface.descriptors() {
+                if let Some(idx) = interface_desc.description_string_index() {
+                    if let Ok(interface_name) = device.read_string_descriptor_ascii(idx) {
+                        let ports = device.port_numbers()?.iter().map(
+                            |id| id.to_string()).collect::<Vec<String>>().join(".");
+                        let interface_path = path
+                            .join(format!("{}-{}", device.bus_number(), ports))
+                            .join(format!("{}-{}:{}.{}",device.bus_number(), ports,
+                                          config_desc.number(), interface.number()));
+                        // Check the ASCII name of this USB interface.
+                        if interface_name == "HyperDebug Shell" {
```

Comment:
This would probably be better as a match; you can do the UART line as `name if name.starts_with("UART") => `.

Reply:
Done